[ADD] mccabe: Add check for mccabe complexity cyclomatic threshold

moylop260 opened the pull request:
- [x] Add new check
- [x] Add rating value in pylint message
- [x] Check if astroid get build a node tree (without compile) [Asked here](https://github.com/PyCQA/pylint/pull/877#discussion_r60832314)
- [x] Add tests


Integrate a new check to verify the [cyclomatic complexity](https://en.wikipedia.org/wiki/Cyclomatic_complexity) of a code with mccabe 

Other python Checker QA [flake8](https://flake8.readthedocs.org/en/latest/#quickstart) have this check.

```bash
flake8 offers an extra option: 
  –max-complexity, which will emit a warning 
       if the McCabe complexity of a function is higher than the value. 
       By default it’s deactivated

$ flake8 --max-complexity 12 coolproject
coolproject/mod.py:939:1: C901 'Checker.check_all' is too complex (12)
coolproject/mod.py:1204:1: C901 'selftest' is too complex (14)
```

moylop260 commented on pylint/extensions/check_mccabe.py:
```diff
@@ -0,0 +1,72 @@
+        """Get a compile python tree of nodes from code"""
+        # TODO: Check if astroid get build a similar tree
+        code = node.as_string()
+        tree = compile(code, '', "exec", ast.PyCF_ONLY_AST)
```
@PCManticore 
Could you help to locate a similar code with astroid or pylint (wihtout compile)?
The `type(tree)` is `<class '_ast.Module'>` and the `code` is `def my_method(): pass`.
I did not found a way to use astroid.

moylop260 commented:
This pr is ready to review

PCManticore commented on pylint/extensions/check_mccabe.py:
```diff
@@ -0,0 +1,71 @@
+        """Get a compile python tree of nodes from code"""
+        # TODO: Check if astroid build a similar tree
+        code = node.as_string()
+        tree = compile(code, '', "exec", ast.PyCF_ONLY_AST)
```
Why do you this? We already have the tree built for us, available through visit_module. Also, if McCabeChecker works only with a builtin ast module, instead of the one provided by astroid, then we should implement the complexity metric ourselves.

moylop260 replied:
@PCManticore Thanks for feedback.

lmedioni commented:
We have been working on this some months ago for a client of us. This could interest you: 

- [complexity checker](https://bitbucket.org/lmedioni/cnes_pylint_extension/commits/69ec0b3c26cffa884934259a6984fc665b1b2389) 

- [simplified complexity](https://bitbucket.org/lmedioni/cnes_pylint_extension/commits/fc435692d9cbae37746741861536b33b827e751f)

We actually reused the [existing maccabe module](https://pypi.python.org/pypi/mccabe) and called it from _visit_module_.

moylop260 commented:
@lmedioni 
Thanks for your contribution I will review it

moylop260 commented:
@PCManticore 
I used mccabe original class with inherit based on @lmedioni dev.
Could you check if is a better approach for you?

PCManticore commented on pylint/extensions/check_mccabe.py:
```diff
@@ -0,0 +1,210 @@
+        self.reset()
+        self._bottom_counter = 0
+
+    def default(self, node, *args):
```
Why does this have to be empty?

PCManticore commented on pylint/extensions/check_mccabe.py:
```diff
@@ -0,0 +1,210 @@
+        tree = compile(code, '', "exec", ast.PyCF_ONLY_AST)
+        return tree
+
+    def _check_too_complex_deprecated(self, node):
```
If this is not used, just remove it.

PCManticore commented on pylint/extensions/check_mccabe.py:
```diff
@@ -0,0 +1,210 @@
+
+    visitAsyncFunctionDef = visitFunctionDef
+
+    def preorder(self, tree, visitor, *args):
```
I don't understand what's happening here, but preorder seems to be called all the time with the same visitor. Why can't this simply be self.dispatch(tree, *args)?

PCManticore commented on pylint/extensions/check_mccabe.py:
```diff
@@ -0,0 +1,210 @@
+        super(McCabeASTVisitor, self).__init__()
+        self.node = None
+        self._cache = {}
+        self.classname = ""
```
Does this ever changes? It seems to be constant across the execution of the program or am I missing something? If it is not used, please simplify the key for the .graphs dictionary.

PCManticore commented on pylint/extensions/check_mccabe.py:
```diff
@@ -0,0 +1,210 @@
+class McCabeASTVisitor(PathGraphingAstVisitor):
+    def __init__(self):
+        super(McCabeASTVisitor, self).__init__()
+        self.node = None
```
Are these satisfying any API requirement from PathGraphingAstVisitor? If not, can they be made private?

PCManticore commented:
In general it looks okay. There are a couple of changes here and there that I want to see before doing another round of reviews.

moylop260 commented:
@PCManticore 
Thanks for feedback
I removed the deprecated code and clean redundant inherit code (code called from super class exactly in inherit class).

moylop260 commented:
Hello @lmedioni 
How I can add you contribution here?

A question:
I added a [file of example](https://github.com/vauxoo-dev/pylint/blob/master-pylint-add-mccabe-moy/pylint/test/extensions/data/mccabe.py)

Running the flake8 script:
```bash
flake8 --select=C901 --max-complexity 1 pylint/test/extensions/data/mccabe.py
# The output was:
pylint/test/extensions/data/mccabe.py:16:1: C901 'f3' is too complex (3)
pylint/test/extensions/data/mccabe.py:26:1: C901 'f4' is too complex (2)
pylint/test/extensions/data/mccabe.py:32:1: C901 'f5' is too complex (2)
pylint/test/extensions/data/mccabe.py:40:1: C901 'f6' is too complex (2)
pylint/test/extensions/data/mccabe.py:48:1: C901 'f7' is too complex (3)
pylint/test/extensions/data/mccabe.py:59:1: C901 'f8' is too complex (4)
pylint/test/extensions/data/mccabe.py:71:1: C901 'f9' is too complex (9)
pylint/test/extensions/data/mccabe.py:97:1: C901 'f10' is too complex (11)
pylint/test/extensions/data/mccabe.py:131:1: C901 'MyClass1.method2' is too complex (18)
pylint/test/extensions/data/mccabe.py:188:1: C901 'Loop 188' is too complex (4)
pylint/test/extensions/data/mccabe.py:197:1: C901 'method3' is too complex (2)
```

Running the pylint with your extension script:
```bash
hg clone ssh://hg@bitbucket.org/lmedioni/cnes_pylint_extension
cd cnes_pylint_extension/checkers
pylint --load-plugins=cnes_checker --max-mccabe-number=1 -r n -d all -e R5301 --output-format=colorized --msg-template="{path}:{line}: [{msg_id}({symbol}), {obj}] {msg}" pylint/test/extensions/data/mccabe.py
# The output was:
************* Module mccabe
pylint/test/extensions/data/mccabe.py:4: [R5301(too-high-complexity), f1] Too high cyclomatic complexity (mccabe 1/0)
pylint/test/extensions/data/mccabe.py:9: [R5301(too-high-complexity), f2] Too high cyclomatic complexity (mccabe 1/0)
pylint/test/extensions/data/mccabe.py:16: [R5301(too-high-complexity), f3] Too high cyclomatic complexity (mccabe 3/0)
pylint/test/extensions/data/mccabe.py:26: [R5301(too-high-complexity), f4] Too high cyclomatic complexity (mccabe 2/0)
pylint/test/extensions/data/mccabe.py:32: [R5301(too-high-complexity), f5] Too high cyclomatic complexity (mccabe 2/0)
pylint/test/extensions/data/mccabe.py:40: [R5301(too-high-complexity), f6] Too high cyclomatic complexity (mccabe 2/0)
pylint/test/extensions/data/mccabe.py:48: [R5301(too-high-complexity), f7] Too high cyclomatic complexity (mccabe 3/0)
pylint/test/extensions/data/mccabe.py:59: [R5301(too-high-complexity), f8] Too high cyclomatic complexity (mccabe 4/0)
pylint/test/extensions/data/mccabe.py:71: [R5301(too-high-complexity), f9] Too high cyclomatic complexity (mccabe 9/0)
pylint/test/extensions/data/mccabe.py:188: [R5301(too-high-complexity), ] Too high cyclomatic complexity (mccabe 4/0)
pylint/test/extensions/data/mccabe.py:127: [R5301(too-high-complexity), MyClass1.method1] Too high cyclomatic complexity (mccabe 1/0)
pylint/test/extensions/data/mccabe.py:97: [R5301(too-high-complexity), f10] Too high cyclomatic complexity (mccabe 11/0)
pylint/test/extensions/data/mccabe.py:131: [R5301(too-high-complexity), MyClass1.method2] Too high cyclomatic complexity (mccabe 12/0)
pylint/test/extensions/data/mccabe.py:197: [R5301(too-high-complexity), method3] Too high cyclomatic complexity (mccabe 1/0)
```

Notice that `pylint/test/extensions/data/mccabe.py:131 method2` show 12 with your script
but show 18 with original flake8.

Do you know what is the problem here?

PCManticore commented on pylint/test/extensions/data/mccabe.py:
```diff
@@ -0,0 +1,206 @@
+
+    def method2(self, param1):
+        """McCabe rating: 18"""
+        # FIXME: Should be the rate of 18 but is 12
```
Don't you have to add a visitFor and a visitTryFinally?

moylop260 replied:
We don't need it

The super class have this visit, we are overwriting just the methods where need a appendNode

To identify them I ran:
`git clone git@github.com:PyCQA/mccabe.git && cd mccabe && cat mccabe.py | grep "    self.appendPathNode" -B 6 | grep "def"`
Output:
```python
def visitSimpleStatement(self, node):
def visitWith(self, node):
```

PCManticore commented on pylint/test/extensions/data/mccabe.py:
```diff
@@ -0,0 +1,206 @@
+            pass
+    finally:
+        pass
+    # FIXME: If you delete the `return` the mccabe increase of 1 to 2 in pylint? (flake8 even is 2)
```
Probably because you are not handling try-finally. There is a special node for that, TryFinally (you only have TryExcept)

PCManticore commented on pylint/test/extensions/test_check_mccabe.py:
```diff
@@ -0,0 +1,61 @@
+        'f10 is too complex. The McCabe rating is 11',
+        'method1 is too complex. The McCabe rating is 1',
+        'method2 is too complex. The McCabe rating is 18',
+        "'For 188' is too complex. The McCabe rating is 4",
```
Where is this test? I can't find any 188 elsewhere.

moylop260 replied:
The test is [pylint/test/extensions/data/mccabe.py#L187](https://github.com/vauxoo-dev/pylint/blob/49991e53f611ddb246766ff0fafd80aeaf255140/pylint/test/extensions/data/mccabe.py#L187)

The code is 
```python
for count in range(10):
    if count == 1:
        exit(0)
    elif count == 2:
        exit(1)
    else:
        exit(2)
```

The four sentence is too complex, but don't have a name (like as methods), then mccabe use: `type sentence` plus line number to identify it
`For 188` is the sentence `for` in the line number 188 that have a mccabe rating of 4 this section (because don't have a parent method node, is a `for` isolated.)

PCManticore commented on pylint/extensions/check_mccabe.py:
```diff
@@ -0,0 +1,176 @@
+                args=(node_name, complexity))
+
+    @check_messages('too-complex')
+    def visit_module(self, node):
```
Isn't this the same as _check_too_complex? Do you plan to use _check_too_complex somewhere else? Otherwise we could just rename _check_too_complex to visit_module.

PCManticore commented on pylint/extensions/check_mccabe.py:
```diff
@@ -0,0 +1,176 @@
+# -*- coding: utf-8 -*-
```
You're using somewhere non-ascii characters? This encoding cookie might be superfluous.

moylop260 commented:
@PCManticore 
Thanks for feedback again.

- I fixed `FIXME` comments https://github.com/PyCQA/pylint/pull/877/commits/8fad072af695186d25416638a96a3c76bbe6a3d7
- Removed superfluous encoding https://github.com/PyCQA/pylint/pull/877/commits/49991e53f611ddb246766ff0fafd80aeaf255140
- Renamed `_check_too_complex` to `visit_module` https://github.com/PyCQA/pylint/pull/877/commits/49991e53f611ddb246766ff0fafd80aeaf255140

moylop260 commented:
@PCManticore 
FYI I tested many existing modules with pylint extension vs flake8 extension and show same result.

Executed:
`python -m pylint --load-plugins=pylint.extensions.check_mccabe --msg-template="{path}:{line}: [{msg_id}({symbol}), {obj}] {msg}" -r n -d all -e R1260 --max-complexity=1 ~/many-modules` <- output [pylint_mccabe.txt](https://github.com/PyCQA/pylint/files/249718/pylint_mccabe.txt)
`flake8 --select=C901 --max-complexity 1 ~/many-modules` <- output [flake8_mccabe.txt](https://github.com/PyCQA/pylint/files/249717/flake8_mccabe.txt)

to detect different result (rate, files...) I have created the following script:
```diff
diff --git a/pylint/extensions/check_mccabe.py b/pylint/extensions/check_mccabe.py
index 7b6d077..94a57b4 100644
--- a/pylint/extensions/check_mccabe.py
+++ b/pylint/extensions/check_mccabe.py
@@ -109,6 +109,21 @@ class PathGraphingAstVisitor(Mccabe_PathGraphingAstVisitor):
             self.tail = bottom
 
 
+def get_fails_expected():
+    import re
+    regex_str = r"(?P<path>[A-Za-z0-9-._/]+):(?P<lineno>\d+):\d+: C901 (?P<sentence>'(.*)'+) is too complex \((?P<rate>\d+)\)"
+    regex = re.compile(regex_str)
+    mydict = {}
+    with open('/Users/moylop260/odoo/pylint/flake8_mccabe.txt') as fp:
+        for line in fp:
+            regex_match = regex.match(line)
+            if regex_match:
+                mydict[(regex_match.group('path'), regex_match.group('lineno'))] = regex_match.group('rate')
+            else:
+                print "line not detected", line
+    return mydict
+
+
 class McCabeMethodChecker(BaseChecker):
     """Checks McCabe complexity cyclomatic threshold in methods and functions
     to validate a too complex code.
@@ -149,9 +164,21 @@ class McCabeMethodChecker(BaseChecker):
                 node_name = "%s %d" % (node.__class__.__name__, node.lineno)
             if complexity <= self.config.max_complexity:
                 continue
+            key = (node.root().path, str(node.lineno))
+            if not key in self.mydict:
+                print "key not found", key
+                import pdb;pdb.set_trace()
+            else:
+                expected_complexity = self.mydict.pop(key)
+                if expected_complexity != str(complexity):
+                    print "%s:%s" % key,
+                    print " expected_complexity %s vs real %s" % (expected_complexity, complexity)
+                    import pdb;pdb.set_trace()
             self.add_message(
                 'too-complex', node=node, confidence=HIGH,
                 args=(node_name, complexity))
+            if len(self.mydict) <= 3:
+                print "self.mydict", self.mydict
 
 
 def register(linter):
@@ -160,4 +187,7 @@ def register(linter):
     :param linter: Main interface object for Pylint plugins
     :type linter: Pylint object
     """
-    linter.register_checker(McCabeMethodChecker(linter))
+    mydict = get_fails_expected()
+    obj = McCabeMethodChecker(linter)
+    obj.mydict = mydict
+    linter.register_checker(obj)
```

And the result was done!

PCManticore commented on pylint/test/extensions/test_check_mccabe.py:
```diff
@@ -0,0 +1,61 @@
+        "'f10' is too complex. The McCabe rating is 11",
+        "'method1' is too complex. The McCabe rating is 1",
+        "'method2' is too complex. The McCabe rating is 18",
+        "'For 187' is too complex. The McCabe rating is 4",
```
We should drop this behaviour, it's not obvious. What can be done instead is:

- for methods, classes just use their name, as we do right now
- for objects without a name, just say what they are, as in "This for is too complex".

moylop260 replied:
> for objects without a name, just say what they are, as in "This for is too complex".

Are you talking about replace the message by: 
`s/'For 187' is too complex. The McCabe rating is 4/This for is too complex. The McCabe rating is 4/g`?

moylop260 commented:
@PCManticore 

- [x] [REF] mccabe: Change message for objects without a name
     - Use "This 'for' is too complex" instead of "'For 187' is too complex"

moylop260 commented:
FYI my last touch... increasing coverage of the file to 100%
![screen shot 2016-05-05 at 12 46 56 pm](https://cloud.githubusercontent.com/assets/6644187/15051685/7cbd804a-12bf-11e6-8029-e4aecb91104a.png)

moylop260 commented:
ping @PCManticore

PCManticore commented:
Thank you!

moylop260 commented:
@PCManticore 
thank for merge and follow up!
😃 👯
